src/device_trezor/device_trezor.cpp

```diff
@@ -69,6 +70,67 @@ namespace trezor {
       }
     }
 
+    bool device_trezor::init()
+    {
+      bool r = device_trezor_base::init();
+      if (r && !m_live_refresh_thread)
+      {
+        m_live_refresh_thread_running = true;
+        m_live_refresh_thread.reset(new boost::thread(boost::bind(&device_trezor::live_refresh_thread_main, this)));
+      }
+      return r;
+    }
+
+    bool device_trezor::release()
+    {
+      bool r = device_trezor_base::release();
+      m_live_refresh_thread_running = false;
+      if (m_live_refresh_thread)
+      {
+        m_live_refresh_thread->join();
+        m_live_refresh_thread.release();
+      }
+      return r;
+    }
+
+    void device_trezor::device_state_reset_unsafe()
+    {
+      require_connected();
+      if (m_live_refresh_in_progress)
+      {
+        try
+        {
+          live_refresh_finish_unsafe();
+        }
+        catch(const std::exception & e)
+        {
+          MERROR("Live refresh could not be terminated: " << e.what());
+        }
+      }
+
+      m_live_refresh_in_progress = false;
+      device_trezor_base::device_state_reset_unsafe();
+    }
+
+    void device_trezor::live_refresh_thread_main()
+    {
+      while(m_live_refresh_thread_running)
+      {
+        boost::this_thread::sleep_for(boost::chrono::milliseconds(10));
+        if (!m_transport || !m_live_refresh_in_progress)
```

Comment:
That seems racy. There is no lock in release, which calls base::release, which calls disconnect, which deletes m_transport and sets it to nullptr.

Reply:
Good catch, I've moved the release after thread join, so it waits for completion.